# **Automatidata project**
**Course 2 - Get Started with Python**

Welcome to the Automatidata Project!

You have just started as a data professional in a fictional data analytics firm, Automatidata. Their client, the New York City Taxi and Limousine Commission (New York City TLC), has hired the Automatidata team for its reputation in helping their clients develop data-based solutions.

The team is still in the early stages of the project. Previously, you were asked to complete a project proposal by your supervisor, DeShawn Washington. You have received notice that your project proposal has been approved and that New York City TLC has given the Automatidata team access to their data. To get clear insights, New York TLC's data must be analyzed, key variables identified, and the dataset ensured it is ready for analysis.

A notebook was structured and prepared to help you in this project. Please complete the questions inside and prepare a summary for the data team.

# Course 2 End-of-course project: Inspect and analyze data

In this activity, you will examine data provided and prepare it for analysis.  
<br/>   

**The purpose** of this project is to investigate and understand the data provided.
  
**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings. 
<br/>  
*This activity has three parts:*

**Part 1:** Understand the situation 
* How can you best prepare to understand and organize the provided taxi cab information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities.

* Compile summary information about the data to inform next steps.

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into specific variables.


<br/> 
Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work. 



# **Identify data types and relevant variables using Python**


### **Exercise Instructions:**


**Complete the following step-by-step instructions to inspect and analyze this NYC taxi dataset.**

This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods

**Follow the instructions and answer questions to complete this activity. Afterward,**

1.   Use the structured notebook provided to help you in this project. Please complete the questions inside and prepare a summary for the data team.

2.  Consider the questions presented  in the [Course 2 PACE strategy document](https://docs.google.com/document/d/1JM7h5MAQkD9uxoUhgpVBeT5R15y85LRaMo30I7bIFdk/template/preview).

3. Write a short Executive Summary using your findings.

4.   Compare your data insights with the provided exemplar to confirm of your approach and results.


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

# **PACE stages** 


<img src="images/Pace.png" width="100" height="100" align=left>

   *        [Plan](#scrollTo=psz51YkZVwtN&line=3&uniqifier=1)
   *        [Analyze](#scrollTo=mA7Mz_SnI8km&line=4&uniqifier=1)
   *        [Construct](#scrollTo=Lca9c8XON8lc&line=2&uniqifier=1)
   *        [Execute](#scrollTo=401PgchTPr4E&line=2&uniqifier=1)

<img src="images/Plan.png" width="100" height="100" align=left>


## PACE: **Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided taxi cab information? 

Read the column descriptions and consider any implications (i.e. driver entered values might be invalid); observe the number of rows and consider the appropriate techniques for handling that amount of data.

<img src="images/Analyze.png" width="100" height="100" align=left>

## PACE: **Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Build dataframe**
















Create a pandas dataframe for data learning, exploratory data analysis (EDA), and statistical activities.

**Code the following,**

*   import pandas as pd               #library exercise for buidling dataframes

*   import numpy as np                #numpy is imported with pandas

*   df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')

**Note:** pair the data object name "df" with pandas functions to manipulate data, such as df.groupby().

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
import numpy as np
import pandas as pd

# RUN THIS CELL TO IMPORT YOUR DATA.
df = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
df.head()

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80

### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by coding the following:

1. df.head(10)
2. df.info()
3. df.describe()

Consider the following two questions:

**Question 1:** When reviewing the df.info() output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 2:** When reviewing the df.describe() output, what do you notice about the distributions of each variable? Are there any questionable values?

1. There are no null values (I used all(~df.isna()) to check this). There are three non-numeric columns: tpep_pickup_datetimem, tpep_dropoff_datetime, and store_and_fwd_flag. All of these are strings. It might be more efficient to case store_and_fwd_flag as a Boolean. I do notice that the range of the index is greater than the number of rows; why are there missing rows? Maybe the missing values were filtered out before I got my hands on this data.

2. It can be easier to understand distributions pictorally, but one thing that stands out is that we should not be performing numerical operations on any of the columns Unnamed: 0, VendorID RatecodeID PULocationID DOLocationID payment_type. There _are_ numbers, yes, but they are ID numbers, they're not meant to participate in arithmetic.

In [6]:
display(df.head())

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [18]:
df.describe()

Unnamed: 0      VendorID  passenger_count  trip_distance  \
count  2.269900e+04  22699.000000     22699.000000   22699.000000   
mean   5.675849e+07      1.556236         1.642319       2.913313   
std    3.274493e+07      0.496838         1.285231       3.653171   
min    1.212700e+04      1.000000         0.000000       0.000000   
25%    2.852056e+07      1.000000         1.000000       0.990000   
50%    5.673150e+07      2.000000         1.000000       1.610000   
75%    8.537452e+07      2.000000         2.000000       3.060000   
max    1.134863e+08      2.000000         6.000000      33.960000   

         RatecodeID  PULocationID  DOLocationID  payment_type   fare_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean       1.043394    162.412353    161.527997      1.336887     13.026629   
std        0.708391     66.633373     70.139691      0.496211     13.243791   
min        1.000000      1.000000      1.000000      1.000000   -120.000000   
25%        1.000000    114.000000    112.000000      1.000000      6.500000   
50%        1.000000    162.000000    162.000000      1.000000      9.500000   
75%        1.000000    233.000000    233.000000      2.000000     14.500000   
max       99.000000    265.000000    265.000000      4.000000    999.990000   

              extra       mta_tax    tip_amount  tolls_amount  \
count  22699.000000  22699.000000  22699.000000  22699.000000   
mean       0.333275      0.497445      1.835781      0.312542   
std        0.463097      0.039465      2.800626      1.399212   
min       -1.000000     -0.500000      0.000000      0.000000   
25%        0.000000      0.500000      0.000000      0.000000   
50%        0.000000      0.500000      1.350000      0.000000   
75%        0.500000      0.500000      2.450000      0.000000   
max        4.500000      0.500000    200.000000     19.100000   

       improvement_surcharge  total_amount  
count           22699.000000  22699.000000  
mean                0.299551     16.310502  
std                 0.015673     16.097295  
min                -0.300000   -120.300000  
25%                 0.300000      8.750000  
50%                 0.300000     11.800000  
75%                 0.300000     17.800000  
max                 0.300000   1200.290000

### **Task 2c. Understand the data - Investigate the variables**

Sort and interpret the data table for two variables:`trip_distance` and `total_amount`.

**Answer the following three questions:**

**Question 1:** Sort your first variable (`trip_distance`) from maximum to minimum value, do the values seem normal?

**Question 2:** Sort your by your second variable (`total_amount`), are any values unusual?

**Question 3:** Are the resulting rows similar for both sorts? Why or why not?

*#==> ENTER YOUR RESPONSES TO QUESTION 1-3 HERE

In [22]:
# Sort the data by trip distance from maximum to minimum value
df["trip_distance"].sort_values()




22026     0.00
795       0.00
6908      0.00
13561     0.00
12238     0.00
         ...  
29       30.83
10291    31.95
6064     32.72
13861    33.92
9280     33.96
Name: trip_distance, Length: 22699, dtype: float64

In [25]:
sum( df["trip_distance"] == 0 )
# Why are there 148 trips of no distance

148

In [29]:
# Sort the data by total amount and print the top 20 values
df["total_amount"].sort_values().tail(20)

# That $1200.29 total is crazy. But I suppose the other values are normal. Maybe that big one was a long limo ride
# or something like that.

29        111.38
13359     111.95
13621     115.94
18130     119.31
7281      120.96
908       121.56
11608     123.30
6708      126.00
10291     131.80
1928      137.80
9280      150.30
11269     151.82
3582      152.30
16379     157.06
6064      179.06
15474     211.80
12511     233.74
13861     258.21
20312     450.30
8476     1200.29
Name: total_amount, dtype: float64

In [28]:
# Sort the data by total amount and print the bottom 20 values
df["total_amount"].sort_values().head(20)
# I don't understand why there are negative values. This is something to ask the TLC about. Also, the number of 0's here 
# is not the same as the number of trips of 0 distance, so those trips are harder to interpret. It would be worth 
# investigating if these negative trips and the trips of 0 distance were voided.

12944   -120.30
20698     -5.80
17602     -5.80
11204     -5.30
14714     -5.30
8204      -4.80
20317     -4.80
10281     -4.30
5448      -4.30
4423      -4.30
18565     -3.80
314       -3.80
5758      -3.80
1646      -3.30
10506      0.00
4402       0.00
5722       0.00
22566      0.00
19067      0.30
14283      0.31
Name: total_amount, dtype: float64

In [30]:
# How many of each payment type are represented in the data?
df["payment_type"].value_counts()

1    15265
2     7267
3      121
4       46
Name: payment_type, dtype: int64

In [37]:
# What is the average tip for trips paid for with credit card?
filt = df["payment_type"] == 1
credit_mean = np.mean(df.loc[filt,"tip_amount"])

# What is the average tip for trips paid for with cash?
filt = df["payment_type"] == 2
cash_mean = np.mean(df.loc[filt,"tip_amount"])

print( "The average tip for trips paid with credit card is $"+str(credit_mean)+".")
print( "The average tip for trips paid with case is $"+str(cash_mean)+". This makes sense because cash tips aren't counted in tip amount.")

The average tip for trips paid with credit card is $2.7298001965279934.
The average tip for trips paid with case is $0.0. This makes sense because cash tips aren't counted in tip amount.


In [38]:
# How many times is each vendor ID represented in the data?
df["VendorID"].value_counts()

2    12626
1    10073
Name: VendorID, dtype: int64

In [40]:
# What is the mean total amount for each vendor?
filt = df["VendorID"] == 1
mean1 = np.mean(df.loc[filt, "total_amount"])
filt = df["VendorID"] == 2
mean2 = np.mean(df.loc[filt, "total_amount"])
print(mean1, mean2)

16.298118733246966 16.32038175193886


In [41]:
# Filter the data for credit card payments only
filt = df["payment_type"] == 1
df.loc[filt]

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0        24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1        35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2       106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3        38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
5        23345809         2   03/25/2017 8:34:11 PM   03/25/2017 8:42:11 PM   
...           ...       ...                     ...                     ...   
22692    60425673         1   07/16/2017 3:22:51 AM   07/16/2017 3:40:52 AM   
22693    67858616         2  08/10/2017 10:20:04 PM  08/10/2017 10:29:31 PM   
22695    66632549         2   08/06/2017 4:43:59 PM   08/06/2017 5:24:47 PM   
22697    60217333         2  07/15/2017 12:56:30 PM   07/15/2017 1:08:26 PM   
22698    17208911         1   03/02/2017 1:02:49 PM   03/02/2017 1:16:09 PM   

       passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                    6           3.34           1                  N   
1                    1           1.80           1                  N   
2                    1           1.00           1                  N   
3                    1           3.70           1                  N   
5                    6           2.30           1                  N   
...                ...            ...         ...                ...   
22692                1           5.70           1                  N   
22693                1           0.89           1                  N   
22695                1          16.71           2                  N   
22697                1           2.36           1                  N   
22698                1           2.10           1                  N   

       PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0               100           231             1         13.0    0.0      0.5   
1               186            43             1         16.0    0.0      0.5   
2               262           236             1          6.5    0.0      0.5   
3               188            97             1         20.5    0.0      0.5   
5               161           236             1          9.0    0.5      0.5   
...             ...           ...           ...          ...    ...      ...   
22692           249            17             1         19.0    0.5      0.5   
22693           229           170             1          7.5    0.5      0.5   
22695           132           164             1         52.0    0.0      0.5   
22697            68           144             1         10.5    0.0      0.5   
22698           239           236             1         11.0    0.0      0.5   

       tip_amount  tolls_amount  improvement_surcharge  total_amount  
0            2.76          0.00                    0.3         16.56  
1            4.00          0.00                    0.3         20.80  
2            1.45          0.00                    0.3          8.75  
3            6.39          0.00                    0.3         27.69  
5            2.06          0.00                    0.3         12.36  
...           ...           ...                    ...           ...  
22692        4.05          0.00                    0.3         24.35  
22693        1.76          0.00                    0.3         10.56  
22695       14.64          5.76                    0.3         73.20  
22697        1.70          0.00                    0.3         13.00  
22698        2.35          0.00                    0.3         14.15  

[15265 rows x 18 columns]

In [42]:
# Filter the data for passenger count only
# Sorry, I can't follow this instruction.

False

In [44]:
# Calculate the average tip amount for each passenger count (credit card payments only)
filt = df["payment_type"] == 1
df.loc[filt].groupby(by="passenger_count").agg(np.mean)["tip_amount"]



passenger_count
0    2.610370
1    2.714681
2    2.829949
3    2.726800
4    2.607753
5    2.762645
6    2.643326
Name: tip_amount, dtype: float64

<img src="images/Construct.png" width="100" height="100" align=left>

## PACE: **Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project. 




<img src="images/Execute.png" width="100" height="100" align=left>

## PACE: **Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response:


### **Given your efforts, what can you summarize for DeShawn and the data team?**

*Note for Learners: Your answer should address Luana's request for a summary that covers the following points:* 

* A summary of the data types of each variable

* The number of null and non-null values

* Preliminary insights derived from the data, including:
    * Unusual or questionable values
    * How many of each payment type are represented in the data
    * Mean tip amount for each payment type
    * How many rides each vendor provided
    * Mean total amount for each vendor
    * Mean tip amount for each passenger count (only of those who paid by credit card)



All of the fields in this data are numeric except for 3: tpep_pickup_datetimem, tpep_dropoff_datetime, and store_and_fwd_flag.

There are no missing values. The total number of rows is 22,699.

Here are a few caveats about this data:
* There are some numeric values which are ID numbers, so those should be excluded from undergoing any mathematical operations.
* There are 148 trips of distance 0. We should find out where these came from.
* There are 14 trips with negative total values. Is this accurate?
* There is one trip whose total cost was over \$1,200. Is this legitimate?

Here is a summary of payment types:
Payment Type    Count

Credit card     15265

Cash            7267

No charge       121

Dispute         46

Unknown         0

Voided          0


Eh, the cells have the summary. I'll type this in Word or something.